File: packages/cithub_native/ios/Classes/Core/WebVpnProtocolClient.swift
import Foundation
#if canImport(Security)
import Security
#endif

public actor WebVpnProtocolClient {
    public static let defaultBaseURL = URL(string: "https://webvpn.ccit.edu.cn/")!
    public static let browserUserAgent =
        "Mozilla/5.0 (iPhone; CPU iPhone OS 18_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Mobile/15E148 Safari/604.1"

    private let store: any SecretStore
    private let logger: any RuntimeLogger
    private let baseURL: URL
    private let transport: any HTTPTransport
    private let cookies: CookieHeaderStore

    public init(
        store: any SecretStore,
        logger: any RuntimeLogger,
        baseURL: URL = WebVpnProtocolClient.defaultBaseURL,
        transport: any HTTPTransport = URLSessionTransport()
    ) {
        self.store = store
        self.logger = logger
        self.baseURL = baseURL.absoluteString.hasSuffix("/") ? baseURL : URL(string: baseURL.absoluteString + "/")!
        self.transport = transport
        self.cookies = CookieHeaderStore(store: store)
    }

    public func loadConfiguration() async throws -> WebVpnLoginConfiguration {
        let data = try await request(method: "GET", path: "api/access/authentication/list").object("data")
        guard let methods = data?.array("list") else {
            throw CithubNativeError.invalidResponse("WebVPN 未返回认证方式")
        }
        for raw in methods {
            guard let method = raw as? [String: Any], method.int64("authType") == 1 else { continue }
            let options = method.object("authOptions")
            let dynamic = options?.array("dynamicVerification")?.compactMap { ($0 as? NSNumber)?.int64Value } ?? []
            return WebVpnLoginConfiguration(
                externalID: method.string("externalId"),
                requiresPassword: options?.int64("staticVerification") == 1,
                requiresCaptcha: options?.int64("useGraphValidateCode") == 1,
                dynamicVerificationTypes: dynamic
            )
        }
        throw CithubNativeError.invalidResponse("学校 WebVPN 当前未开放本地账号登录")
    }

    public func loadCaptcha() async throws -> WebVpnCaptcha {
        guard let data = try await request(
            method: "GET",
            path: "api/access/graph-captcha/validate-code?width=150&height=50"
        ).object("data") else {
            throw CithubNativeError.invalidResponse("WebVPN 接口未返回验证码")
        }
        let id = data.string("id")
        let image = data.string("captcha")
        guard !id.isEmpty, !image.isEmpty else {
            throw CithubNativeError.invalidResponse("WebVPN 验证码响应不完整")
        }
        return WebVpnCaptcha(id: id, image: image)
    }

    public func login(
        username: String,
        password: String,
        captchaID: String,
        captchaCode: String,
        configuration: WebVpnLoginConfiguration
    ) async throws -> WebVpnUser {
        guard configuration.dynamicVerificationTypes.isEmpty else {
            throw CithubNativeError.loginRequired("学校当前要求动态验证码，请先使用官方 WebVPN 网页完成登录")
        }
        var payload: [String: Any] = [
            "deviceId": try await deviceID(),
            "userName": username,
        ]
        if configuration.requiresPassword {
            payload["password"] = try encryptPassword(password)
        }
        if configuration.requiresCaptcha {
            payload["captchaId"] = captchaID
            payload["code"] = captchaCode
        }
        let payloadData = try JSONSerialization.data(withJSONObject: payload)
        let payloadString = String(decoding: payloadData, as: UTF8.self)
        _ = try await request(
            method: "POST",
            path: "api/access/auth/finish",
            json: ["externalId": configuration.externalID, "data": payloadString]
        )
        return try await userInfo()
    }

    public func userInfo() async throws -> WebVpnUser {
        guard let data = try await request(method: "GET", path: "api/access/user/info").object("data") else {
            throw CithubNativeError.invalidResponse("WebVPN 接口未返回用户信息")
        }
        let action: CoreRequiredAction
        if data.bool("needToBindLocalAccount") { action = .bindAccount }
        else if data.bool("needTriggerTFA") { action = .tfa }
        else if data.bool("needChangePwd") { action = .passwordReset }
        else { action = .none }
        let groups = data.array("groups")?.compactMap { $0 as? String } ?? []
        return WebVpnUser(
            username: data.string("username"),
            nickname: data.string("nickname"),
            fullName: data.string("fullName"),
            groups: groups,
            authType: data.int64("authType"),
            bindWechat: data.bool("bindWechat"),
            bindOtp: data.bool("bindOtp"),
            requiredAction: action
        )
    }

    public func logout() async {
        _ = try? await request(method: "POST", path: "api/access/user/logout", json: [:])
        await clearSession()
    }

    public func hasRestorableSession() async -> Bool { !(await cookies.header()).isEmpty }
    public func cookieHeader() async -> String { await cookies.header() }
    public func clearSession() async { await cookies.clear() }

    private func request(method: String, path: String, json: [String: Any]? = nil) async throws -> [String: Any] {
        guard let url = URL(string: path, relativeTo: baseURL)?.absoluteURL else {
            throw CithubNativeError.invalidInput("无效 WebVPN 请求地址")
        }
        var headers = [
            "Accept": "application/json, text/plain, */*",
            "Accept-Language": "zh-CN,zh;q=0.9",
            "User-Agent": Self.browserUserAgent,
            "Origin": baseURL.absoluteString.trimmingCharacters(in: CharacterSet(charactersIn: "/")),
            "Referer": baseURL.absoluteString,
            "DNT": "1",
        ]
        let cookie = await cookies.header()
        if !cookie.isEmpty { headers["Cookie"] = cookie }
        var body = Data()
        if let json {
            headers["Content-Type"] = "application/json"
            body = try JSONSerialization.data(withJSONObject: json)
        }
        let response = try await transport.perform(HTTPCall(method: method, url: url, headers: headers, body: body))
        if let setCookie = response.header("Set-Cookie") { await cookies.merge([setCookie]) }
        await logger.append(source: "webvpn", message: "\(method) /\(path) -> HTTP \(response.status)")
        guard (200...299).contains(response.status) else {
            let message = (try? jsonObject(response.data).string("message")) ?? ""
            throw CithubNativeError.requestFailed(message.isEmpty ? "WebVPN 请求失败（HTTP \(response.status)）" : message)
        }
        let envelope = try jsonObject(response.data)
        let code = envelope.int64("code", default: -1)
        guard code == 0 else {
            let message = envelope.string("message", default: "WebVPN 接口返回错误：\(code)")
            throw CithubNativeError.requestFailed(message)
        }
        return envelope
    }

    private func deviceID() async throws -> String {
        if let existing = await store.get("webvpn.device_id"),
           existing.range(of: "^[0-9a-f]{32}$", options: .regularExpression) != nil {
            return existing
        }
        let value = UUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased()
        await store.put("webvpn.device_id", value)
        return value
    }

    private func encryptPassword(_ password: String) throws -> String {
        #if canImport(Security)
        let clean = Self.publicKeyPEM
            .replacingOccurrences(of: "-----BEGIN PUBLIC KEY-----", with: "")
            .replacingOccurrences(of: "-----END PUBLIC KEY-----", with: "")
            .components(separatedBy: .whitespacesAndNewlines).joined()
        guard let keyData = Data(base64Encoded: clean) else {
            throw CithubNativeError.invalidResponse("WebVPN 公钥格式无效")
        }
        let attributes: [CFString: Any] = [
            kSecAttrKeyType: kSecAttrKeyTypeRSA,
            kSecAttrKeyClass: kSecAttrKeyClassPublic,
            kSecAttrKeySizeInBits: 2048,
        ]
        var keyError: Unmanaged<CFError>?
        guard let key = SecKeyCreateWithData(keyData as CFData, attributes as CFDictionary, &keyError) else {
            throw CithubNativeError.invalidResponse("WebVPN 公钥无法载入")
        }
        guard SecKeyIsAlgorithmSupported(key, .encrypt, .rsaEncryptionPKCS1) else {
            throw CithubNativeError.invalidResponse("当前系统不支持 WebVPN RSA 加密")
        }
        var encryptError: Unmanaged<CFError>?
        guard let encrypted = SecKeyCreateEncryptedData(
            key, .rsaEncryptionPKCS1, Data(password.utf8) as CFData, &encryptError
        ) else {
            throw CithubNativeError.requestFailed("WebVPN 密码加密失败")
        }
        return (encrypted as Data).base64EncodedString()
        #else
        throw CithubNativeError.requestFailed("当前平台不支持 WebVPN RSA 加密")
        #endif
    }

    private static let publicKeyPEM = """
    -----BEGIN PUBLIC KEY-----
    MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAvrqdXbn6tf2kabHLRoE9
    IASO5fZixKK5IsFcBMJ0h1tf0WUb3HMygcC3+NecScetMSoPmSOrDLSA6sBWwGEF
    LTefRM5vP/eFdkXXB0YpFjfganpBKv4ZOvzCWZGhHOUlACRHViazsZbaPHvLYhsH
    Z3XTSbS8iIVDYgrQCHgzs2ULWEUau3489HTAcg7A2V2ZfDDzqaHj5BU5vopbfmjs
    cXObP0Ddy4IW4Mc/fcJoJs1e7M4hZg6iTIb8OTnlssOikckenO9mV+GdxdOSG9K2
    lUTCS+qxFXQ/vgd7JWi0eTOYG2duEoA2u2T3b/G5I/h8En+tOG6Ax0rztp/YtF0Q
    zQIDAQAB
    -----END PUBLIC KEY-----
    """
}

File: packages/cithub_native/android/src/main/kotlin/com/aquasofts/cithub_flutter/WebVpnProtocolClient.kt
package com.aquasofts.cithub_flutter

import com.aquasofts.cithub_flutter.native.RequiredAccountAction
import com.aquasofts.cithub_flutter.native.UserInfoDto
import java.net.HttpURLConnection
import java.net.URL
import java.security.KeyFactory
import java.security.SecureRandom
import java.security.spec.X509EncodedKeySpec
import java.util.Base64
import javax.crypto.Cipher
import org.json.JSONObject

internal data class WebVpnLoginConfiguration(
    val externalId: String,
    val requiresPassword: Boolean,
    val requiresCaptcha: Boolean,
    val dynamicVerificationTypes: List<Int>,
)

internal data class WebVpnCaptcha(val id: String, val image: String)

internal data class WebVpnUser(
    val dto: UserInfoDto,
    val requiredAction: RequiredAccountAction,
)

internal class WebVpnProtocolClient(
    private val store: SecretStore,
    private val logs: RuntimeLogStore,
    baseUrl: String = BASE_URL,
) {
    private val baseUrl = baseUrl.ensureTrailingSlash()
    @Volatile private var cookieHeader: String = store.get(COOKIE_KEY).orEmpty()

    fun loadConfiguration(): WebVpnLoginConfiguration {
        val data = request("GET", "api/access/authentication/list").data()
        val methods = data.optJSONArray("list") ?: error("WebVPN 未返回认证方式")
        for (index in 0 until methods.length()) {
            val method = methods.getJSONObject(index)
            if (method.optInt("authType") != 1) continue
            val options = method.optJSONObject("authOptions")
            val dynamic = options?.optJSONArray("dynamicVerification")
            return WebVpnLoginConfiguration(
                method.getString("externalId"),
                options?.optInt("staticVerification") == 1,
                options?.optInt("useGraphValidateCode") == 1,
                if (dynamic == null) emptyList() else (0 until dynamic.length()).map(dynamic::optInt),
            )
        }
        error("学校 WebVPN 当前未开放本地账号登录")
    }

    fun loadCaptcha(): WebVpnCaptcha {
        val data = request("GET", "api/access/graph-captcha/validate-code?width=150&height=50").data()
        return WebVpnCaptcha(data.getString("id"), data.getString("captcha"))
    }

    fun login(
        username: String,
        password: String,
        captchaId: String,
        captchaCode: String,
        configuration: WebVpnLoginConfiguration,
    ): WebVpnUser {
        require(configuration.dynamicVerificationTypes.isEmpty()) {
            "学校当前要求动态验证码，请先使用官方 WebVPN 网页完成登录"
        }
        val payload = JSONObject()
            .put("deviceId", deviceId())
            .put("userName", username)
        if (configuration.requiresPassword) payload.put("password", encryptPassword(password))
        if (configuration.requiresCaptcha) {
            payload.put("captchaId", captchaId)
            payload.put("code", captchaCode)
        }
        val body = JSONObject()
            .put("externalId", configuration.externalId)
            .put("data", payload.toString())
        request("POST", "api/access/auth/finish", body)
        return userInfo()
    }

    fun userInfo(): WebVpnUser = parseUser(request("GET", "api/access/user/info").data())

    fun logout() {
        runCatching { request("POST", "api/access/user/logout", JSONObject()) }
        clearSession()
    }

    fun hasRestorableSession(): Boolean = cookieHeader.isNotBlank()

    fun cookies(): String = cookieHeader

    fun clearSession() {
        cookieHeader = ""
        store.remove(COOKIE_KEY)
    }

    private fun request(method: String, path: String, json: JSONObject? = null): JSONObject {
        val connection = URL(baseUrl + path).openConnection() as HttpURLConnection
        connection.requestMethod = method
        connection.connectTimeout = 20_000
        connection.readTimeout = 20_000
        connection.instanceFollowRedirects = false
        connection.setRequestProperty("Accept", "application/json, text/plain, */*")
        connection.setRequestProperty("Accept-Language", "zh-CN,zh;q=0.9")
        connection.setRequestProperty("User-Agent", BROWSER_USER_AGENT)
        connection.setRequestProperty("Origin", baseUrl.removeSuffix("/"))
        connection.setRequestProperty("Referer", baseUrl)
        connection.setRequestProperty("DNT", "1")
        if (cookieHeader.isNotBlank()) connection.setRequestProperty("Cookie", cookieHeader)
        if (json != null) {
            connection.doOutput = true
            connection.setRequestProperty("Content-Type", "application/json")
            connection.outputStream.use { it.write(json.toString().toByteArray()) }
        }
        val status = connection.responseCode
        mergeCookies(connection.headerFields.entries
            .firstOrNull { it.key?.equals("set-cookie", ignoreCase = true) == true }
            ?.value.orEmpty())
        val stream = if (status in 200..299) connection.inputStream else connection.errorStream
        val text = stream?.bufferedReader()?.use { it.readText() }.orEmpty()
        logs.append("webvpn", "$method /$path -> HTTP $status")
        if (status !in 200..299) {
            val message = runCatching { JSONObject(text).optString("message") }.getOrNull()
            error(message?.takeIf(String::isNotBlank) ?: "WebVPN 请求失败（HTTP $status）")
        }
        val envelope = JSONObject(text)
        val code = envelope.optInt("code", -1)
        if (code != 0) error(envelope.optString("message", "WebVPN 接口返回错误：$code"))
        return envelope
    }

    @Synchronized
    private fun mergeCookies(setCookies: List<String>) {
        if (setCookies.isEmpty()) return
        val cookies = linkedMapOf<String, String>()
        cookieHeader.split(';').map(String::trim).filter { it.contains('=') }.forEach { pair ->
            cookies[pair.substringBefore('=')] = pair.substringAfter('=')
        }
        setCookies.forEach { raw ->
            val pair = raw.substringBefore(';').trim()
            if (pair.contains('=')) {
                val name = pair.substringBefore('=')
                val value = pair.substringAfter('=')
                if (value.isBlank() || raw.contains("Max-Age=0", ignoreCase = true)) cookies.remove(name)
                else cookies[name] = value
            }
        }
        cookieHeader = cookies.entries.joinToString("; ") { "${it.key}=${it.value}" }
        if (cookieHeader.isBlank()) store.remove(COOKIE_KEY) else store.put(COOKIE_KEY, cookieHeader)
    }

    private fun JSONObject.data(): JSONObject = optJSONObject("data")
        ?: error("WebVPN 接口未返回数据")

    private fun parseUser(data: JSONObject): WebVpnUser {
        val groupsJson = data.optJSONArray("groups")
        val groups = if (groupsJson == null) emptyList() else (0 until groupsJson.length()).map(groupsJson::optString)
        val action = when {
            data.optBoolean("needToBindLocalAccount") -> RequiredAccountAction.BIND_ACCOUNT
            data.optBoolean("needTriggerTFA") -> RequiredAccountAction.TFA
            data.optBoolean("needChangePwd") -> RequiredAccountAction.PASSWORD_RESET
            else -> RequiredAccountAction.NONE
        }
        return WebVpnUser(
            UserInfoDto(
                data.optString("username"),
                data.optString("nickname"),
                data.optString("fullName"),
                groups,
                data.optLong("authType"),
                data.optBoolean("bindWechat"),
                data.optBoolean("bindOtp"),
            ),
            action,
        )
    }

    private fun deviceId(): String {
        store.get(DEVICE_ID_KEY)?.takeIf { it.matches(Regex("^[0-9a-f]{32}$")) }?.let { return it }
        val bytes = ByteArray(16).also(SecureRandom()::nextBytes)
        return bytes.joinToString("") { "%02x".format(it) }.also { store.put(DEVICE_ID_KEY, it) }
    }

    private fun encryptPassword(password: String): String {
        val publicKeyBytes = PUBLIC_KEY_PEM
            .replace("-----BEGIN PUBLIC KEY-----", "")
            .replace("-----END PUBLIC KEY-----", "")
            .replace("\\s".toRegex(), "")
            .let(Base64.getDecoder()::decode)
        val publicKey = KeyFactory.getInstance("RSA").generatePublic(X509EncodedKeySpec(publicKeyBytes))
        return Cipher.getInstance("RSA/ECB/PKCS1Padding").run {
            init(Cipher.ENCRYPT_MODE, publicKey)
            Base64.getEncoder().encodeToString(doFinal(password.toByteArray()))
        }
    }

    private companion object {
        const val BASE_URL = "https://webvpn.ccit.edu.cn/"
        const val COOKIE_KEY = "webvpn.session.cookies"
        const val DEVICE_ID_KEY = "webvpn.device_id"
        const val BROWSER_USER_AGENT = "Mozilla/5.0 (Linux; Android 15) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/150.0.0.0 Mobile Safari/537.36"
        const val PUBLIC_KEY_PEM = """
-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAvrqdXbn6tf2kabHLRoE9
IASO5fZixKK5IsFcBMJ0h1tf0WUb3HMygcC3+NecScetMSoPmSOrDLSA6sBWwGEF
LTefRM5vP/eFdkXXB0YpFjfganpBKv4ZOvzCWZGhHOUlACRHViazsZbaPHvLYhsH
Z3XTSbS8iIVDYgrQCHgzs2ULWEUau3489HTAcg7A2V2ZfDDzqaHj5BU5vopbfmjs
cXObP0Ddy4IW4Mc/fcJoJs1e7M4hZg6iTIb8OTnlssOikckenO9mV+GdxdOSG9K2
lUTCS+qxFXQ/vgd7JWi0eTOYG2duEoA2u2T3b/G5I/h8En+tOG6Ax0rztp/YtF0Q
zQIDAQAB
-----END PUBLIC KEY-----
"""
    }
}

private fun String.ensureTrailingSlash(): String = if (endsWith('/')) this else "$this/"
